cartographer/mapping/trajectory_connectivity_state.cc

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2017 The Cartographer Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cartographer/mapping/trajectory_connectivity_state.h"
+
+namespace cartographer {
+namespace mapping {
+
+TrajectoryConnectivityState::TrajectoryConnectivityState()
+    : connected_components_(),
+      last_connection_time_map_() {}
+
+void TrajectoryConnectivityState::Add(int trajectory_id) {
+  connected_components_.Add(trajectory_id);
+}
+
+void TrajectoryConnectivityState::Connect(int trajectory_id_a,
+                                          int trajectory_id_b,
+                                          const common::Time& time) {
+  if (TransitivelyConnected(trajectory_id_a, trajectory_id_b)) {
+    // The trajectories are transitively connected, i.e. they belong to the same
+    // connected component. In this case we only update the last connection time
+    // of those two trajectories.
+    auto sorted_pair = std::minmax(trajectory_id_a, trajectory_id_b);
+    if (last_connection_time_map_[sorted_pair] < time) {
+      last_connection_time_map_[sorted_pair] = time;
+    }
+  } else {
+    // The connection between these two trajectories is about to join to
```

Comment:
Instead of describing what we do here, describe why: we want to quickly change to more efficient loop closure searches when connected components are joined.

Reply:
Done